### PR TITLE
T14027

### PR DESCRIPTION
**Pull request by wgreenberg**

### CODE REVIEW

File: src/eos-shard-shard-file.c [hunk 1]
```diff
@@ -398,6 +396,23 @@ eos_shard_shard_file_list_records (EosShardShardFile *self)
...
+ * Calls function on each record
+ *
+ */
+void
+eos_shard_shard_file_records_foreach (EosShardShardFile *self,
+                                      EosShardShardFileRecordsForeachFunc func)
+{
+  EosShardShardFileImplInterface *iface = EOS_SHARD_SHARD_FILE_IMPL_GET_IFACE (self->impl);
+  iface->records_foreach (self->impl, func);
+  return;
```
Comment (magcius) [inline src/eos-shard-shard-file.c]: Needless return.

File: src/eos-shard-shard-file.c [hunk 2]
```diff
@@ -398,6 +396,23 @@ eos_shard_shard_file_list_records (EosShardShardFile *self)
   return iface->list_records (self->impl);
 }
 
+/**
+ * eos_shard_shard_file_records_foreach:
+ * @self: the file
+ * @func: (scope call): foo
```
Comment (magcius) [inline src/eos-shard-shard-file.c]: Leave blank if you really don't have better docs.

File: src/eos-shard-shard-file.c [hunk 3]
```diff
@@ -398,6 +396,23 @@ eos_shard_shard_file_list_records (EosShardShardFile *self)
...
+ *
+ * Calls function on each record
+ *
+ */
+void
+eos_shard_shard_file_records_foreach (EosShardShardFile *self,
+                                      EosShardShardFileRecordsForeachFunc func)
```
Comment (magcius) [inline src/eos-shard-shard-file.c]: Use GFunc and accept a user_data.

File: src/eos-shard-shard-file.c [hunk 4]
```diff
@@ -32,8 +32,6 @@
 #include "eos-shard-record.h"
 #include "eos-shard-dictionary.h"
 
-#include "eos-shard-shard-file-impl.h"
```
Comment (magcius) [inline src/eos-shard-shard-file.c]: why did you remove this?

Comment (wgreenberg) [inline src/eos-shard-shard-file.c]: moved it to the header for the function pointer reference

File: src/eos-shard-shard-file.c (hunk 2 above)
Comment (wgreenberg) [inline src/eos-shard-shard-file.c]: oop, meant to replace

File: src/eos-shard-shard-file.c (hunk 3 above)
Comment (wgreenberg) [inline src/eos-shard-shard-file.c]: yup already wrote it up, just have to push

**Comment by wgreenberg** [comment]
@magcius revised!

File: src/eos-shard-shard-file-impl-v2.c [hunk 5]
```diff
@@ -250,6 +250,23 @@ list_records (EosShardShardFileImpl *impl)
   return l;
 }
 
+static void
+records_foreach (EosShardShardFileImpl *impl, EosShardRecordsForeachFunc func, gpointer user_data)
+{
+  EosShardShardFileImplV2 *self = EOS_SHARD_SHARD_FILE_IMPL_V2 (impl);
+  int i;
+
+  for (i=0; i<self->hdr.records_length; i++) {
```
Comment (magcius) [inline src/eos-shard-shard-file-impl-v2.c]: Nit: `i = 0; i < self->hdr.records_length`

File: src/eos-shard-shard-file.c [hunk 6]
```diff
@@ -398,6 +396,24 @@ eos_shard_shard_file_list_records (EosShardShardFile *self)
...
 }
 
+/**
+ * eos_shard_shard_file_records_foreach:
+ * @self: the file
+ * @func: (scope call): the function to call for each record
+ * @user_data: (allow-none): user data to pass to the function
+ *
+ * Calls function on each record
+ *
```
Comment (magcius) [inline src/eos-shard-shard-file.c]: Nit: extra line

**Comment by wgreenberg** [comment]
revised the revision
